builtin/logical/database/backend.go

```diff
@@ -286,3 +485,333 @@ cassandra, mssql, mysql, postgres
+	}
+
+	// Get the Database object
+	db, err := b.GetConnection(ctx, s, input.Role.DBName)
+	if err != nil {
+		return output, err
+	}
+
+	db.RLock()
+	defer db.RUnlock()
+
+	// Use password from input if available. This happens if we're restoring from
+	// a WAL item or processing the rotation queue with an item that has a WAL
+	// associated with it
+	newPassword := input.Password
+	if newPassword == "" {
+		// Generate a new password
+		newPassword, err = db.GenerateCredentials(ctx)
+		if err != nil {
+			return output, err
+		}
+	}
+	output.Password = newPassword
+
+	config := dbplugin.StaticUserConfig{
+		Username: input.Role.StaticAccount.Username,
+		Password: newPassword,
+	}
+
+	if output.WALID == "" {
+		output.WALID, err = framework.PutWAL(ctx, s, staticWALKey, &setCredentialsWAL{
+			RoleName:          input.RoleName,
+			Username:          config.Username,
+			NewPassword:       config.Password,
+			OldPassword:       input.Role.StaticAccount.Password,
+			Statements:        input.Role.Statements,
+			LastVaultRotation: input.Role.StaticAccount.LastVaultRotation,
+		})
+		if err != nil {
+			return output, errwrap.Wrapf("error writing WAL entry: {{err}}", err)
+		}
+	}
+
+	_, password, err := db.SetCredentials(ctx, input.Role.Statements, config)
+	if err != nil {
+		b.CloseIfShutdown(db, err)
+		return output, errwrap.Wrapf("error setting credentials: {{err}}", err)
+	}
+
+	if newPassword != password {
+		return output, errors.New("mismatch passwords returned")
+	}
+
+	// Store updated role information
+	lvr = time.Now()
+	input.Role.StaticAccount.LastVaultRotation = lvr
+	input.Role.StaticAccount.Password = password
+	output.RotationTime = lvr
+
+	entry, err := logical.StorageEntryJSON(databaseStaticRolePath+input.RoleName, input.Role)
+	if err != nil {
+		return output, err
+	}
+	if err := s.Put(ctx, entry); err != nil {
+		return output, err
+	}
+
+	// cleanup WAL after successfully rotating and pushing new item on to queue
+	if err := framework.DeleteWAL(ctx, s, output.WALID); err != nil {
+		merr = multierror.Append(merr, err)
+		return output, merr
+	}
+
+	// the WAL has been deleted, return new setStaticAccountOutput without it
+	return &setStaticAccountOutput{RotationTime: lvr}, merr
+}
+
+// pushItem wraps the internal queue's PushItem call, to make sure a queue is
+// actually available. This is needed because both runTicker and initQueue
+// operate in go-routines, and could be accessing the queue concurrently
+func (b *databaseBackend) pushItem(item *queue.Item) error {
+	b.RLock()
+	unlockFunc := b.RUnlock
+	defer func() { unlockFunc() }()
+
+	if b.credRotationQueue != nil {
+		return b.credRotationQueue.Push(item)
+	}
+	// Upgrade lock
+	b.RUnlock()
+	b.Lock()
+	unlockFunc = b.Unlock
+
+	// check again
+	if b.credRotationQueue != nil {
+		return b.credRotationQueue.Push(item)
+	}
+	b.credRotationQueue = queue.New()
+
+	return b.credRotationQueue.Push(item)
+}
```

Comment:
After all the comments on this code are addressed (since i don't know if github will move the comments over) what do you think about moving all the queue/rotation functions added to this file into their own file? Something like rotation.go or similar.